Don't Re-Shrink PDFs

willp240 opened the pull request:
This PR modifies the logic in BinnedEDManager to not re-shrink a PDF if it's normalisation is not handled as INDIRECT. 
To illustrate the problem, I first committed an update to the unit tests. So if you checkout the first commit, the new test will fail. If you then checkout the second commit, the test (and all others) pass.

The issue manifested itself when you use a buffer, the shrunk pdf would be re-shrunk. On the second shrinking, the new PDF size is now `OriginalSize-2*buffer_region` (where `buffer_region` is sum of high and low buffers). So in the shrinking process, only that many bins get assigned values. Then when it comes to calculate the LLH for `OriginalSize-1*buffer_region` bins you get zero probability bins

The fix just uses the if statement from pre #25, but only for the Direct/False normalisation cases. There was a fixme comment before that if statement, saying have more obvious behaviour. But I think it's ok. Happy to hear suggestions though

Finally, the formatting is a bit off in BinnedEDManager.cpp. Somehow it looks like it evaded my previous formatting changes, so I'll fix that in another PR

dcookman commented:
Okay, I've finally gotten round to looking at this PR again.
As a first step, I've been able to confirm that the code here does pass all unit tests, including the new one you added, and the existing code in `master` fails that unit test.
Now to try and work out what Past Me spotted about this code...(why did I not write it down!)

dcookman reviewed (changes requested):
I've worked out where some of the bugs are again!
So the bugs in this code (including the one you've already fixed in this PR) arise if you use the shrinker **without a systematic**. When the SystematicManager is run over the PDFs, it will always reset the size of the `BinnedED` objects within `fWorkingPdfs` back to their unshrunk number of bins. That's why I got away with the line of code:
```
fWorkingPdfs[i] = shrinker_.ShrinkDist(fWorkingPdfs.at(i));
```
However, as you showed in your unit test, if you only run the shrinker without any systematics, it all goes horribly wrong because there's nothing to reset the PDF bin size every iteration.
Your fix in this PR does kinda work, but it only partially solves the problem. Here are some of the problems I spotted:

- The bug is still present for the third, `INDIRECT` case: you can check this by creating two more identical unit tests to the one you added here, but setting the `NormFittingStatus`in them to `INDIRECT` and `FALSE`. The `FALSE` one will pass now, but the `INDIRECT` one still crashes.
- You only check the 0th indexed PDF each time to see whether it's been shrunk or not.
- Moreover, if the 0th PDF has already been shrunk, you immediately return from the whole method! Is that the functionality we want? (Maybe?)
- In line 178, we exit early if the `fWorkingPdfs` vector is empty. The claim in the comment above is that we should "only shrink if not already shrunk", but this check **does not do that.** That's because we add a PDF to the `fWorkingPdfs` vector during line 129 of the `BinnedEDManager::AddPdf()` method. So in fact all this line does is to check if no PDFs have been added at all! Which is not a huge problem, but it's completely redundant (the rest of the method ends immediately if the vector is empty anyway), and the comment is misleading. We should either remove the check entirely, or come up with some other way for correctly flagging if all PDFs have been shrunk.
- Looking at the structure, it occurs to me that there's a bunch of redundant lines of code in this method. We might be able to re-jig things a bit.

willp240 commented:
Thanks @dcookman! Your comments made this much easier for me to get my head around and I think now I've got a proper solution. I've added similar tests for each of the fitting statuses, and have restructured the if statements on fitting status to have slightly less repeated code I think.

The actual fix for the original problem is a check on the number of bins in the WorkingPDF. If this equals the number in the OriginalPDF minus the buffer, I assume the shrinking has already been applied (probably because the user has a buffer but no systematics) and so we just return out the method. Otherwise, if the number of bins in the WorkingPDF doesn't equal the number in the OriginalPDF we throw as something has gone horribly wrong. I've written this in a comment in the code. I feel when a long comment like that is needed maybe it suggests the code should be clearer, but it's better than no comment at all! 

I've also removed the check that there are some WorkingPDFs and the misleading comment with it

So overall, other than a slight restructuring not much functionality changes for the majority of use cases, but we're now robust to that weird edge case of using a buffer with no systematics. I've removed the not ready to merge label as this can be reviewed now

dcookman reviewed (approved):
Lovely! This all seems a lot better. Okay there is still technically some real weird edge cases where this could break I guess, but at that point you're doing a really silly kind of analysis.
It compiles for me and passes all unit tests - thanks for the new ones btw.
Ready to merge in (let's get you antinu guys back onto the master branch!).
